src/main/scala/circt/stage/phases/CIRCT.scala

```diff
@@ -0,0 +1,171 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package circt.stage.phases
+
+import circt.Implicits.BooleanImplicits
+import circt.stage.{CIRCTOptions, CIRCTTarget, EmittedMLIR, PreserveAggregate}
+
+import firrtl.{AnnotationSeq, EmittedVerilogCircuit, EmittedVerilogCircuitAnnotation}
+import firrtl.options.{
+  CustomFileEmission,
+  Dependency,
+  OptionsException,
+  OutputAnnotationFileAnnotation,
+  Phase,
+  StageError,
+  StageOptions,
+  StageUtils
+}
+import firrtl.options.phases.WriteOutputAnnotations
+import firrtl.options.Viewer.view
+import firrtl.stage.{FirrtlOptions, RunFirrtlTransformAnnotation}
+import _root_.logger.LogLevel
+
+import scala.sys.process._
+
+private object Helpers {
+  implicit class LogLevelHelpers(logLevel: LogLevel.Value) {
+    def toCIRCTOptions: Seq[String] = logLevel match {
+      case LogLevel.Error => Seq.empty
+      case LogLevel.Warn  => Seq.empty
+      case LogLevel.Info  => Seq("-verbose-pass-executions")
+      case LogLevel.Debug => Seq("-verbose-pass-executions")
+      case LogLevel.Trace => Seq("-verbose-pass-executions", "-print-ir-after-all")
+      case LogLevel.None  => Seq.empty
+    }
+  }
+}
+
+/** A phase that calls and runs CIRCT, specifically `firtool`, while preserving an [[AnnotationSeq]] API.
+  *
+  * This is analogous to [[firrtl.stage.phases.Compiler]].
+  */
+class CIRCT extends Phase {
+
+  import Helpers._
+
+  override def prerequisites = Seq(
+    Dependency[firrtl.stage.phases.AddDefaults],
+    Dependency[firrtl.stage.phases.AddImplicitEmitter],
+    Dependency[firrtl.stage.phases.AddImplicitOutputFile]
+  )
+  override def optionalPrerequisites = Seq.empty
+  override def optionalPrerequisiteOf = Seq.empty
+  override def invalidates(a: Phase) = false
+
+  override def transform(annotations: AnnotationSeq): AnnotationSeq = {
+    val circtOptions = view[CIRCTOptions](annotations)
+    val firrtlOptions = view[FirrtlOptions](annotations)
+    val stageOptions = view[StageOptions](annotations)
+
+    var blackbox, inferReadWrite = false
+    var dedup, imcp = true
+    var logLevel = _root_.logger.LogLevel.None
+
+    val annotationsx = annotations.filter {
+      case _: CustomFileEmission           => false
+      case _: firrtl.EmitCircuitAnnotation => false
+      case RunFirrtlTransformAnnotation(transform) =>
+        transform match {
+          /* Inlining/Flattening happen by default, so these can be dropped. */
+          case _: firrtl.passes.InlineInstances | _: firrtl.transforms.Flatten => false
+          /* ReplSeqMem is converted to a firtool option */
+          case _: firrtl.passes.memlib.ReplSeqMem =>
+            blackbox = true
+            false
+          /* Any emitters should not be passed to firtool. */
+          case _: firrtl.Emitter => false
```

Comment:
How do you specify the target if you emitters are ignored? Like how would I distinguish between generating Firrtl or SystemVerilog?

Reply:
There's a `CIRCTTargetAnnotation` that will control it. They main issue here is that there isn't a good way of adding multiple `RunFirrtlTransformAnnotation(_: firrtl.Emitter)`s to CIRCT as it's trying to work towards a fixed target, e.g., FIRRTL Dialect, HW Dialect, SV Dialect, or SV.

It's reasonable that if there was one `RunFirrtlTransformAnnotation(_ : firrtl.Emitter)` that this could be used to tell CIRCT what to compile to. However, it seemed cleaner to just strip this and force users to use the new annotation.

Reply:
Thinking about this a bit more... it would probably be better to actually error on this. I can do this in a follow-up. 

This should be basically dead code right now as the expectation is that a user is going to use `circt.stage.ChiselStage` which is _not_ going to use a `RunFirrtlTransformAnnotation` to tell CIRCT what emitter to use. There's a separate API/Annotation for that, `CIRCTTargetAnnotation`. A user could see weird behavior if they are doing one of the following:
1. Generating a `RunFirrtlTransformAnnotation(_ : Emitter)` inside a `ChiselAnnotation`.
2. Building a custom stage out of Chisel3 phases and CIRCT phases. 
3. Possibly other weird things.